File: Tar/Python/Test/General/MergeSort.py
def merge_sort(data):  # O(n log n)
    if len(data) <= 1:
        return
    mid = int(len(data) / 2)
    left_data = data[:mid]
    right_data = data[mid:]
    merge_sort(left_data)
    merge_sort(right_data)
    left_index = 0
    right_index = 0
    data_index = 0
    while (left_index < len(left_data) and right_index < len(right_data)):
        if left_data[left_index] < right_data[right_index]:
            data[data_index] = left_data[left_index]
            left_index += 1
        else:
            data[data_index] = right_data[right_index]
            right_index += 1
        data_index += 1
    if left_index < len(left_data):
        print("left index less")
        print(data)
        del data[data_index:]
        data += left_data[left_index:]
    elif right_index < len(right_data):
        print("right index less")
        print(data)
        del data[data_index:]
        data += right_data[right_index:]

if __name__ == '__main__':
    data = [-22, 1, 0, 1, 4, 2, 0, 0, 5, 65]
    merge_sort(data)
    #print(data)

File: Tar/C++/AlgosImpl/Sorting/MergeSort.cpp
#include<bits/stdc++.h>
using namespace std;

void mergeSort(vector<int> &arr) {
    if (arr.size() <= 1) {
        return;
    }
    int n = arr.size();
    int mid = n / 2;
    vector<int> leftData;
    vector<int> rightData;
    for (int i = 0; i < mid; ++i) {
        leftData.push_back(arr.at(i));
    }
    for (int i = mid; i < n; ++i) {
        rightData.push_back(arr.at(i));
    }
    mergeSort(leftData);
    mergeSort(rightData);
    int leftIndex = 0;
    int rightIndex = 0;
    int dataIndex = 0;
    int leftDataSz = leftData.size();
    int rightDataSz = rightData.size();
    while (leftIndex < leftDataSz && rightIndex < rightDataSz) {
        if (leftData.at(leftIndex) < rightData.at(rightIndex)) {
            arr.at(dataIndex) = leftData.at(leftIndex);
            leftIndex += 1;
        } else {
            arr.at(dataIndex) = rightData.at(rightIndex);
            rightIndex += 1;
        }
        dataIndex += 1;
    }
    if (leftIndex < leftDataSz) {
        arr.erase(arr.begin() + dataIndex, arr.begin() + n);
        for (int i = leftIndex; i < leftDataSz; ++i) {
            arr.push_back(leftData.at(i));
        }
    } else if (rightIndex < rightDataSz) {
        arr.erase(arr.begin() + dataIndex, arr.begin() + n);
        for (int i = rightIndex; i < rightDataSz; ++i) {
            arr.push_back(rightData.at(i));
        }
    }
}


int main() {
    vector<int> arr = { -22, 1, 0, 3, 4, 8, 5};
    mergeSort(arr);
    for (int num : arr) {
        cout << num << " ";
    }
}